### PR TITLE
operations: fix possible unintended --size-only skip with multithread…

### LINKED ISSUE #6206: Make rclone more robust to false-negatives with local copies and --size-only
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/does-size-only-account-for-local-sparse-files-properly/31022/5

#### What is your current rclone version (output from `rclone version`)?

```
rclone v1.58.1
- os/version: darwin 10.15.7 (64 bit)
- os/kernel: 19.6.0 (x86_64)
- os/type: darwin
- os/arch: amd64
- go/version: go1.17.9
- go/linking: dynamic
- go/tags: cmount
```

#### What problem are you are trying to solve?

(theoretical I think but...) if doing a copy from a remote to local with multiple streams and `--size-only` (either explicit or implied by the remote, e.g., WebDAV), it is conceivable for an interrupted transfer to not get restarted since it's "size" is correct even if the content is not

#### How do you think rclone should be changed to solve that?

Either add an extra null byte that gets truncated or withhold the very last byte until the rest of the transfer is complete. This way, the size of the file always mismatches until completed

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

### PR DESCRIPTION
**Pull request by Evan-sp**
… copy

#### What is the purpose of this change?

Fix #6206 by withholding the very last byte until the rest of the transfer is complete

#### Was the change discussed in an issue or in the forum before?

#6206

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by Jwink3101** [comment]
Is there an update on this? Plan to merge?

File: fs/operations/multithread.go [hunk 1]
```diff
@@ -178,10 +179,25 @@ func multiThreadCopy(ctx context.Context, f fs.Fs, remote string, src fs.Object,
 		})
 	}
 	err = g.Wait()
-	closeErr := mc.wc.Close()
 	if err != nil {
 		return nil, err
 	}
+
+	// Copy last withheld byte
```
Comment (ncw) [inline fs/operations/multithread.go:186]: This Open will cause another HTTP transaction to the backend and hence quite a lot of extra work / delay for each file.

Ideally what we'd do is fetch the last byte above where we have the connection open, stash it away, and store it here.

**Review by ncw — COMMENTED** [review]
I think this is a great idea.

The implementation could be made more efficient though - please see inline.

**Comment by Jwink3101** [comment]
I mentioned in the original forum post that an alternative would be to preallocate an extra byte and truncate. Since this is _only_ for local destination, can all local remotes support truncate? That would solve the book-keeping and be simple.

**Comment by ncw** [comment]
Yes that seems like a good idea.

You should be able to truncate the file with OpenWriterAt by opening it again I think.

We could always add truncate to that interface if necessary.

**Comment by Jwink3101** [comment]
Just checking if there has been any progress on this? It is a conceivable scenario and it worries me that we aren't protected. (e.g., downloading WebDAV to local).

@ncw? @Evan-sp

**Comment by ncw** [comment]
I'd like to get the efficiency thing resolved, either by keeping the last byte, or using the preallocate trick.

@Evan-sp do you want to have a go at this? If not how about you @Jwink3101 ?

**Comment by ncw** [comment]
In 5c594fea90429ccc35328a3c387b9d326167cfa9 we implemented transfers to a temporary file name on the local, ftp and sftp backends unless the `--inplace` flag is not used.

I think this therefore fixes the issue that this PR is intended to fix as the file won't be renamed to its proper name unless the download completes properly.

I'm therefore going to close this PR.

Thank you @Evan-sp for doing it and sorry we didn't get it merged.

**Pull request closed without merging**